### DISCUSSION TITLE
Future of UII? Public roadmap?

### DISCUSSION THREAD

**Discussion by reactima** [epam/UUI#851] [Ideas]
Hi UUI team, 

First of all, thank you for all your hard work! UUI repo became my top destination and reference source for how to write very good, clean and modern Reactjs/TS code. 

From my understanding UUI is in all major EPAM projects and realise notes reflect it. 
https://uui.epam.com/documents?id=releaseNotes&mode=doc&skin=UUI4_promo

- But what is the future of UUI? 
- Do you have any future roadmap for UUI?
- How do you compare UUI with other major React UI kits? 

Hope it's not too much to ask ... 

It would really helpful if @jakobz as Solution Architect or any core team member will get some publicity and tell more on Youtube or english or russian speaking IT media. 

Your work 100% deserved more publicity!

Here's some links to reference for comparison. 

- Tailwind 
	- https://tailwindui.com/changelog
	- https://github.com/tailwindlabs/headlessui 
	- https://tailwindui.com/components

- Microsoft Fluent UI React 
	- https://react.fluentui.dev/?path=/docs/concepts-introduction--page
	- Fluent UI React https://react.fluentui.dev/?path=/docs/concepts-introduction--page
	- Fluent UI React Roadmap https://github.com/microsoft/fluentui/wiki/Fluent-UI-React-Roadmap
	- Hooks, Charts, etc everything https://github.com/microsoft/fluentui/tree/master/packages/
	- Unification https://github.com/microsoft/fluentui/wiki/FAQ---Fabric-and-Stardust-to-Fluent-UI
	- Editor https://fluentsite.z22.web.core.windows.net/0.64.0/shorthand-props
	- Table https://react.fluentui.dev/?path=/docs/preview-components-table--default
	
* Alibaba ANT
	- https://ant.design/  
	- https://ant.design/components/overview/
	- https://antv.vision/en
	- https://pro.ant.design/docs/upgrade-v5/

* Blueprint 
	- https://blueprintjs.com/docs/
	- https://github.com/palantir/blueprint/wiki/Frequently-Asked-Questions
 
Feel free to reach me out if you want to hear an outsider opinion ;) 

telegram: @reactima 
whatsapp +(81)80 4756 2323

**Comment by jakobz** [comment]
Hi!

Sorry for late answers, and thanks a lot for suggestions and references!

We plan to put more efforts in documentation, education, and publicity next year. There's a plan to make a tech-talk on UUI roadmaps and plans. 

> Do you have any future roadmap for UUI?

As a quick overview, for 2023 the plan is:
- complete 'themes' support: we are working on eliminating code duplication between our 'skins', by moving all styles differences to CSS (css variables). As a 2nd goal, we'll get an ability to customize UUI design to match customer's brand books - thus making way simpler UUI adoption for customers. The first release is planned in Q1 2023
- face-lift and update most components - we have a huge list of minor improvements, which are post-poned due to "themes" migration
- React 18 support
- migration of RTE to plate.js (started, ETA Q1 2023)
- extend Data Tables are related components features: add/remove/move rows, excel-like cell replication, etc. 
- UUI site redesign
- more documentation, examples, education materials

> How do you compare UUI with other major React UI kits?

We have an deck with such comparison, and it's a good idea to put this to uui.epam.com. In few words - UUI is focused on business apps, and is more feature-complete for such kind of apps: we have Data Tables, Rich Text Editor, Forms handling - all bundled and styled. Libraries like MUI are for more generic usage, and usually limited to set of simpler components. Ant Design is a closest competitor, yet UUI is already more feature-complete in many areas.

> But what is the future of UUI?

We feel like stopping racing for more features, and put more efforts in documentation, visibility, quality, accessibility. We already working on making UUI visible for EPAM customer-facing teams, and we have 3 customer projects using UUI. So the ultimate goal is to compete with other sol Ant Design on global market, and as an accelerator for apps EPAM builds for our Customers.

**Reply by reactima** [reply]
@jakobz thank you for such detailed reply! Really appreciate it! 

Sadly I decided to opt out for Ant Design already. Found Antd quite manageable after 5.0 rewrite which React 18 ready hooks and very interesting approach to theme styling and under hood css-in-js hashing 
https://ant.design/docs/blog/css-in-js

UUI repo is a great place to be inspired. Ex, Api Context and Error Handling in UUI gives me ideas how to code UI in "enterprise-ready" fashion 

https://uui.epam.com/documents?id=apiContext&mode=doc&skin=UUI4_promo&category=contexts

Ant has another great advantage for me - it wont break Japanese, Chinese and other asian, double byte languages localisations as it's obviously written and actively maintained by Alibaba team who mostly use Chinese in their day work.

> migration of RTE to plate.js (started, ETA Q1 2023)

I've also played with plate.js and found its "ui" package a bit annoying as it pulls twin.macro, tailwind and few other nasty deps for all examples 
https://github.com/udecode/plate/tree/main/packages/ui 

Building editor with plate's headless repo seems require a full time efforts. Nothing easy.

It would be very interesting to see how you and your team will interate on plate/slate.js for document editing. So far UUI Rich Text Editor looks very decent.
https://uui.epam.com/demo?id=RTE

What I would really love to see in the future is "enterprise-ready" UUI framework which wont break things and dont pull too many repos and legacy hacks, and it seems your roadmap is aligned with it. Will be happy to revisit UUI in the future!

P.s. Please have a look "End-to-end typesafe APIs" concept. It's quite powerful and potentially unlocks UI front generation right from backend data scheme. 
https://trpc.io/
